In [6]:
import pandas as pd
import numpy as np

import warnings
warnings.filterwarnings(action='ignore')
import tensorflow as tf

import math
import matplotlib.pyplot as plt

from keras.models import Sequential
from keras.layers import Dense, LSTM, Dropout
from sklearn.preprocessing import MinMaxScaler
from sklearn.metrics import mean_squared_error
import pickle
import gzip
import tensorflow as tf

gpus = tf.config.experimental.list_physical_devices('GPU')
print(gpus)
if gpus:
    try:
        for i in range(len(gpus)):
            tf.config.experimental.set_memory_growth(gpus[i], True)
    except RuntimeError as e:
        # 프로그램 시작시에 메모리 증가가 설정되어야만 합니다
        print(e)

[PhysicalDevice(name='/physical_device:GPU:0', device_type='GPU'), PhysicalDevice(name='/physical_device:GPU:1', device_type='GPU')]


In [7]:
import keras
keras.__version__

'2.4.3'

# Data Load

In [19]:
import random  
seed_num = 42
random.seed(seed_num)

x = np.load('/project/LSH/x_(7727,10,4068).npy')
y = np.load('/project/LSH/y_(7727,1).npy')

idx = list(range(len(x)))
random.shuffle(idx)

i = round(x.shape[0]*0.8)
X_train, y_train = x[idx[:i],:,:], y[idx[:i]]
X_test, y_test = x[idx[i:],:,:], y[idx[i:]]

X_train.shape, y_train.shape, X_test.shape, y_test.shape

((6182, 10, 4068), (6182,), (1545, 10, 4068), (1545,))

In [9]:
def MAE(y_true, y_pred):
    return "{:.2e}".format(np.mean(np.abs(y_true-y_pred)))

def MSE(y_true, y_pred):
    return "{:.2e}".format(np.mean(np.square(y_true-y_pred)))

def RMSE(y_true, y_pred):
    return "{:.2e}".format(np.sqrt(np.mean(np.square(y_true-y_pred))))

In [10]:
from sklearn.ensemble import AdaBoostRegressor, AdaBoostClassifier
from keras.wrappers.scikit_learn import KerasClassifier, KerasRegressor
from sklearn.ensemble import VotingClassifier
from keras.layers import InputLayer, GRU

# Modeling

In [7]:
# # 2. LSTM model
# def lstm_model():
#     seed_num = 42 # model_seed
#     tf.random.set_seed(seed_num)

#     lstm = Sequential()
#     lstm.add(InputLayer(input_shape=(x.shape[1],x.shape[2])))
#     lstm.add(LSTM(units=128, activation='hard_sigmoid', return_sequences=True))
#     lstm.add(LSTM(units=64, activation='hard_sigmoid', return_sequences=True))
#     lstm.add(Dropout(0.2))
#     lstm.add(LSTM(units=64, activation='hard_sigmoid', return_sequences=True))
#     lstm.add(LSTM(units=32, activation='hard_sigmoid', return_sequences=False))
#     lstm.add(Dropout(0.2))
#     lstm.add(Dense(units=1, activation='sigmoid'))
#     return lstm

# def sorted_ls(path):
#     mtime=lambda f: os.stat(os.path.join(path, f)).st_mtime
#     return list(sorted(os.listdir(path), key=mtime))

# path = '/project/LSH/'
# with tf.device('/device:GPU:0'):
#     # 1. Data load
#     x = np.load(path + 'x_(7727,10,4068).npy')
#     y = np.load(path + 'y_(7727,1).npy')

#     # 3-1. Best model saving
#     MODEL_SAVE_FOLDER_PATH = './models/'
#     if not os.path.exists(MODEL_SAVE_FOLDER_PATH):
#         os.mkdir(MODEL_SAVE_FOLDER_PATH)

#     acc_list, precision_list, recall_list, f1_list, auc_list = [], [], [], [], []
#     sss = StratifiedShuffleSplit(n_splits=10, test_size = 0.2, random_state = 42)

#     # 4. Crossvalidation
#     for seed, (train_index, test_index) in tqdm(enumerate(sss.split(x, y))):
#         X_train, y_train = x[train_index,:,:], y[train_index]
#         X_test, y_test = x[test_index,:,:], y[test_index]

#         # 3-2. Best model saving
#         model_path = MODEL_SAVE_FOLDER_PATH + f'/seed={seed}_'+'{epoch:02d}'+'_val_loss:{val_loss:.4f}.hdf5'

#         cb_checkpoint = ModelCheckpoint(filepath=model_path, monitor='val_loss',
#                                         verbose=1, save_best_only=True)
#         early_stop = EarlyStopping(monitor='val_loss', patience=50, verbose=1, restore_best_weights=False)


#         model = lstm_model()
#         model.compile(optimizer= keras.optimizers.Adam(learning_rate = 0.001), 
#                       loss = "binary_crossentropy", metrics=['acc'])
#         model.fit(X_train, y_train, validation_split=0.25, batch_size=128, epochs=500,
#                   callbacks=[early_stop, cb_checkpoint], shuffle=False)


In [11]:
tf.random.set_seed(seed_num)

In [18]:
%%time
tf.random.set_seed(seed_num)
def get_model():
    lstm = Sequential()
    lstm.add(InputLayer(input_shape=(x.shape[1],x.shape[2])))
    lstm.add(LSTM(units=128, activation='hard_sigmoid', return_sequences=True))
    lstm.add(LSTM(units=64, activation='hard_sigmoid', return_sequences=True))
    lstm.add(Dropout(0.2))
    lstm.add(LSTM(units=64, activation='hard_sigmoid', return_sequences=True))
    lstm.add(LSTM(units=32, activation='hard_sigmoid', return_sequences=False))
    lstm.add(Dropout(0.2))
    lstm.add(Dense(units=1, activation='sigmoid'))

    lstm.compile(optimizer= keras.optimizers.Adam(learning_rate = 0.001), 
                          loss = "binary_crossentropy", metrics=['acc'])
    return lstm

estimator = []
for i in range(1,51):
    LSTM_Predictors = KerasClassifier(build_fn=lambda:get_model(), epochs=20, batch_size=256)
    LSTM_Predictors._estimator_type="classifier"
    estimator.append((f'model{i}', LSTM_Predictors))
print(estimator)    

# estimator = [('model1', LSTM_Predictors), ('model2', LSTM_Predictors1)]

final_model = VotingClassifier(estimators = estimator, voting = 'soft')
final_model.fit(X_train,y_train)

preds = final_model.predict(X_test)

preds[preds>0.5]=1
preds[preds<=0.5]=0

from sklearn import metrics 
print('정확도 :', metrics.accuracy_score(y_test, preds))
print("RMSE : ", RMSE(y_test, preds))

[('model1', <tensorflow.python.keras.wrappers.scikit_learn.KerasClassifier object at 0x7fa32029f850>), ('model2', <tensorflow.python.keras.wrappers.scikit_learn.KerasClassifier object at 0x7fa35c0a2af0>), ('model3', <tensorflow.python.keras.wrappers.scikit_learn.KerasClassifier object at 0x7fa35c0a2e20>), ('model4', <tensorflow.python.keras.wrappers.scikit_learn.KerasClassifier object at 0x7fa3187ed0d0>), ('model5', <tensorflow.python.keras.wrappers.scikit_learn.KerasClassifier object at 0x7fa3205f4430>), ('model6', <tensorflow.python.keras.wrappers.scikit_learn.KerasClassifier object at 0x7fa3205f4e80>), ('model7', <tensorflow.python.keras.wrappers.scikit_learn.KerasClassifier object at 0x7fa3205f4c40>), ('model8', <tensorflow.python.keras.wrappers.scikit_learn.KerasClassifier object at 0x7fa3205f4ee0>), ('model9', <tensorflow.python.keras.wrappers.scikit_learn.KerasClassifier object at 0x7fa3205f4d60>), ('model10', <tensorflow.python.keras.wrappers.scikit_learn.KerasClassifier object

KeyboardInterrupt: 

**KerasClassifier(build_fn=lambda:get_model(), epochs=20, batch_size=256)**
- range(1,2) : 0.747572
- range(1,11) : 0.7508090614886731
- range(1,6) : 0.7462783171521036
- range(1,21) : 0.7540453074433657
- 50개 : 0.7501618122977346

In [38]:
%%time
tf.random.set_seed(seed_num)
def get_model():
    lstm = Sequential()
    lstm.add(InputLayer(input_shape=(x.shape[1],x.shape[2])))
    lstm.add(LSTM(units=128, activation='hard_sigmoid', return_sequences=True))
    lstm.add(LSTM(units=64, activation='hard_sigmoid', return_sequences=True))
    lstm.add(Dropout(0.2))
    lstm.add(LSTM(units=64, activation='hard_sigmoid', return_sequences=True))
    lstm.add(LSTM(units=32, activation='hard_sigmoid', return_sequences=False))
    lstm.add(Dropout(0.2))
    lstm.add(Dense(units=1, activation='sigmoid'))

    lstm.compile(optimizer= keras.optimizers.Adam(learning_rate = 0.001), 
                          loss = "binary_crossentropy", metrics=['acc'])
    return lstm
with tf.device('/device:GPU:0'):
    estimator = []
    for i in range(1,51):
        LSTM_Predictors = KerasClassifier(build_fn=lambda:get_model(), epochs=100, batch_size=256)
        LSTM_Predictors._estimator_type="classifier"
        estimator.append((f'model{i}', LSTM_Predictors))
    print(estimator)    

    # estimator = [('model1', LSTM_Predictors), ('model2', LSTM_Predictors1)]

    final_model = VotingClassifier(estimators = estimator, voting = 'soft')
    final_model.fit(X_train,y_train)

    preds = final_model.predict(X_test)

    preds[preds>0.5]=1
    preds[preds<=0.5]=0

    from sklearn import metrics 
    print('정확도 :', metrics.accuracy_score(y_test, preds))
    print("RMSE : ", RMSE(y_test, preds))

[('model1', <tensorflow.python.keras.wrappers.scikit_learn.KerasClassifier object at 0x7fa320269460>), ('model2', <tensorflow.python.keras.wrappers.scikit_learn.KerasClassifier object at 0x7fa3e6ba08b0>), ('model3', <tensorflow.python.keras.wrappers.scikit_learn.KerasClassifier object at 0x7fa3e6ba0e80>), ('model4', <tensorflow.python.keras.wrappers.scikit_learn.KerasClassifier object at 0x7fa3e6ba0e50>), ('model5', <tensorflow.python.keras.wrappers.scikit_learn.KerasClassifier object at 0x7fa3e6ba0b50>), ('model6', <tensorflow.python.keras.wrappers.scikit_learn.KerasClassifier object at 0x7fa3e6ba0370>), ('model7', <tensorflow.python.keras.wrappers.scikit_learn.KerasClassifier object at 0x7fa3e6ba0fd0>), ('model8', <tensorflow.python.keras.wrappers.scikit_learn.KerasClassifier object at 0x7fa3e6ba0280>), ('model9', <tensorflow.python.keras.wrappers.scikit_learn.KerasClassifier object at 0x7fa3e6ba07f0>), ('model10', <tensorflow.python.keras.wrappers.scikit_learn.KerasClassifier object

Epoch 26/100
25/25 [==============================] - 2s 68ms/step - loss: 0.1367 - acc: 0.9609
Epoch 27/100
25/25 [==============================] - 2s 67ms/step - loss: 0.1405 - acc: 0.9615
Epoch 28/100
25/25 [==============================] - 2s 68ms/step - loss: 0.1270 - acc: 0.9655
Epoch 29/100
25/25 [==============================] - 2s 66ms/step - loss: 0.1275 - acc: 0.9646
Epoch 30/100
25/25 [==============================] - 2s 67ms/step - loss: 0.1244 - acc: 0.9660
Epoch 31/100
25/25 [==============================] - 2s 67ms/step - loss: 0.1240 - acc: 0.9665
Epoch 32/100
25/25 [==============================] - 2s 67ms/step - loss: 0.1240 - acc: 0.9662
Epoch 33/100
25/25 [==============================] - 2s 67ms/step - loss: 0.1185 - acc: 0.9681
Epoch 34/100
25/25 [==============================] - 2s 67ms/step - loss: 0.1118 - acc: 0.9717
Epoch 35/100
25/25 [==============================] - 2s 67ms/step - loss: 0.1040 - acc: 0.9749
Epoch 36/100
25/25 [====================

25/25 [==============================] - 2s 67ms/step - loss: 0.6082 - acc: 0.6679
Epoch 5/100
25/25 [==============================] - 2s 68ms/step - loss: 0.5103 - acc: 0.7676
Epoch 6/100
25/25 [==============================] - 2s 68ms/step - loss: 0.4395 - acc: 0.8067
Epoch 7/100
25/25 [==============================] - 2s 68ms/step - loss: 0.4152 - acc: 0.8261
Epoch 8/100
25/25 [==============================] - 2s 68ms/step - loss: 0.3716 - acc: 0.8499
Epoch 9/100
25/25 [==============================] - 2s 68ms/step - loss: 0.3447 - acc: 0.8682
Epoch 10/100
25/25 [==============================] - 2s 67ms/step - loss: 0.3248 - acc: 0.8792
Epoch 11/100
25/25 [==============================] - 2s 67ms/step - loss: 0.3097 - acc: 0.8885
Epoch 12/100
25/25 [==============================] - 2s 67ms/step - loss: 0.2866 - acc: 0.9029
Epoch 13/100
25/25 [==============================] - 2s 68ms/step - loss: 0.2676 - acc: 0.9097
Epoch 14/100
25/25 [==============================] - 2s 6

25/25 [==============================] - 2s 68ms/step - loss: 0.0876 - acc: 0.9795
Epoch 89/100
25/25 [==============================] - 2s 69ms/step - loss: 0.0808 - acc: 0.9825
Epoch 90/100
25/25 [==============================] - 2s 69ms/step - loss: 0.0793 - acc: 0.9832
Epoch 91/100
25/25 [==============================] - 2s 69ms/step - loss: 0.0793 - acc: 0.9838
Epoch 92/100
25/25 [==============================] - 2s 68ms/step - loss: 0.0899 - acc: 0.9782
Epoch 93/100
25/25 [==============================] - 2s 67ms/step - loss: 0.0797 - acc: 0.9837
Epoch 94/100
25/25 [==============================] - 2s 68ms/step - loss: 0.0786 - acc: 0.9832
Epoch 95/100
25/25 [==============================] - 2s 68ms/step - loss: 0.0776 - acc: 0.9837
Epoch 96/100
25/25 [==============================] - 2s 68ms/step - loss: 0.0762 - acc: 0.9837
Epoch 97/100
25/25 [==============================] - 2s 68ms/step - loss: 0.0777 - acc: 0.9837
Epoch 98/100
25/25 [==============================] -

25/25 [==============================] - 2s 67ms/step - loss: 0.1020 - acc: 0.9770
Epoch 67/100
25/25 [==============================] - 2s 69ms/step - loss: 0.1051 - acc: 0.9761
Epoch 68/100
25/25 [==============================] - 2s 68ms/step - loss: 0.1025 - acc: 0.9770
Epoch 69/100
25/25 [==============================] - 2s 68ms/step - loss: 0.0989 - acc: 0.9793
Epoch 70/100
25/25 [==============================] - 2s 68ms/step - loss: 0.0992 - acc: 0.9785
Epoch 71/100
25/25 [==============================] - 2s 68ms/step - loss: 0.1031 - acc: 0.9759
Epoch 72/100
25/25 [==============================] - 2s 69ms/step - loss: 0.1008 - acc: 0.9777
Epoch 73/100
25/25 [==============================] - 2s 67ms/step - loss: 0.1054 - acc: 0.9749
Epoch 74/100
25/25 [==============================] - 2s 67ms/step - loss: 0.0995 - acc: 0.9767
Epoch 75/100
25/25 [==============================] - 2s 68ms/step - loss: 0.0979 - acc: 0.9793
Epoch 76/100
25/25 [==============================] -

25/25 [==============================] - 2s 67ms/step - loss: 0.0940 - acc: 0.9762
Epoch 45/100
25/25 [==============================] - 2s 67ms/step - loss: 0.0964 - acc: 0.9743
Epoch 46/100
25/25 [==============================] - 2s 68ms/step - loss: 0.0839 - acc: 0.9811
Epoch 47/100
25/25 [==============================] - 2s 67ms/step - loss: 0.0808 - acc: 0.9825
Epoch 48/100
25/25 [==============================] - 2s 67ms/step - loss: 0.0819 - acc: 0.9799
Epoch 49/100
25/25 [==============================] - 2s 67ms/step - loss: 0.0772 - acc: 0.9822
Epoch 50/100
25/25 [==============================] - 2s 67ms/step - loss: 0.0767 - acc: 0.9825
Epoch 51/100
25/25 [==============================] - 2s 66ms/step - loss: 0.0809 - acc: 0.9809
Epoch 52/100
25/25 [==============================] - 2s 67ms/step - loss: 0.0767 - acc: 0.9817
Epoch 53/100
25/25 [==============================] - 2s 67ms/step - loss: 0.0768 - acc: 0.9829
Epoch 54/100
25/25 [==============================] -

25/25 [==============================] - 2s 67ms/step - loss: 0.1736 - acc: 0.9473: 0s - loss: 0.1614 - acc: 0. - ETA: 0s - loss: 0.1599 - a
Epoch 23/100
25/25 [==============================] - 2s 67ms/step - loss: 0.1620 - acc: 0.9520
Epoch 24/100
25/25 [==============================] - 2s 67ms/step - loss: 0.1489 - acc: 0.9609: 0s - loss: 0.141
Epoch 25/100
25/25 [==============================] - 2s 67ms/step - loss: 0.1396 - acc: 0.9638
Epoch 26/100
25/25 [==============================] - 2s 67ms/step - loss: 0.1393 - acc: 0.9636: 1s - loss: 0
Epoch 27/100
25/25 [==============================] - 2s 69ms/step - loss: 0.1332 - acc: 0.9646
Epoch 28/100
25/25 [==============================] - 2s 67ms/step - loss: 0.1314 - acc: 0.9630
Epoch 29/100
25/25 [==============================] - 2s 68ms/step - loss: 0.1369 - acc: 0.9599
Epoch 30/100
25/25 [==============================] - 2s 67ms/step - loss: 0.1327 - acc: 0.9638
Epoch 31/100
25/25 [==============================] - 2s 68

Epoch 1/100
25/25 [==============================] - 5s 66ms/step - loss: 0.6901 - acc: 0.5611
Epoch 2/100
25/25 [==============================] - 2s 66ms/step - loss: 0.6732 - acc: 0.6060
Epoch 3/100
25/25 [==============================] - 2s 66ms/step - loss: 0.6693 - acc: 0.6074
Epoch 4/100
25/25 [==============================] - 2s 67ms/step - loss: 0.6575 - acc: 0.6131
Epoch 5/100
25/25 [==============================] - 2s 67ms/step - loss: 0.5946 - acc: 0.6773
Epoch 6/100
25/25 [==============================] - 2s 67ms/step - loss: 0.4893 - acc: 0.7850
Epoch 7/100
25/25 [==============================] - 2s 67ms/step - loss: 0.4463 - acc: 0.8111: 0s - loss: 0.4467 - acc: 0.8
Epoch 8/100
25/25 [==============================] - 2s 67ms/step - loss: 0.3987 - acc: 0.8382: 0s - loss: 0.3982 - acc: 0.83
Epoch 9/100
25/25 [==============================] - 2s 68ms/step - loss: 0.3666 - acc: 0.8599
Epoch 10/100
25/25 [==============================] - 2s 67ms/step - loss: 0.3445 - 

Epoch 82/100
25/25 [==============================] - 2s 67ms/step - loss: 0.0965 - acc: 0.9795
Epoch 83/100
25/25 [==============================] - 2s 67ms/step - loss: 0.0910 - acc: 0.9809
Epoch 84/100
25/25 [==============================] - 2s 68ms/step - loss: 0.0952 - acc: 0.9790
Epoch 85/100
25/25 [==============================] - 2s 68ms/step - loss: 0.1029 - acc: 0.9751
Epoch 86/100
25/25 [==============================] - 2s 68ms/step - loss: 0.1169 - acc: 0.9701
Epoch 87/100
25/25 [==============================] - 2s 67ms/step - loss: 0.0968 - acc: 0.9790
Epoch 88/100
25/25 [==============================] - 2s 68ms/step - loss: 0.0912 - acc: 0.9804
Epoch 89/100
25/25 [==============================] - 2s 68ms/step - loss: 0.0892 - acc: 0.9812
Epoch 90/100
25/25 [==============================] - 2s 68ms/step - loss: 0.0884 - acc: 0.9816
Epoch 91/100
25/25 [==============================] - 2s 67ms/step - loss: 0.0897 - acc: 0.9804
Epoch 92/100
25/25 [====================

25/25 [==============================] - 2s 67ms/step - loss: 0.0864 - acc: 0.9796
Epoch 61/100
25/25 [==============================] - 2s 68ms/step - loss: 0.0943 - acc: 0.9785
Epoch 62/100
25/25 [==============================] - 2s 67ms/step - loss: 0.0901 - acc: 0.9790
Epoch 63/100
25/25 [==============================] - 2s 66ms/step - loss: 0.0968 - acc: 0.9762
Epoch 64/100
25/25 [==============================] - 2s 67ms/step - loss: 0.0840 - acc: 0.9814
Epoch 65/100
25/25 [==============================] - 2s 67ms/step - loss: 0.0839 - acc: 0.9811
Epoch 66/100
25/25 [==============================] - 2s 67ms/step - loss: 0.0853 - acc: 0.9811
Epoch 67/100
25/25 [==============================] - 2s 67ms/step - loss: 0.0825 - acc: 0.9817
Epoch 68/100
25/25 [==============================] - 2s 68ms/step - loss: 0.0838 - acc: 0.9814
Epoch 69/100
25/25 [==============================] - 2s 68ms/step - loss: 0.0836 - acc: 0.9816
Epoch 70/100
25/25 [==============================] -

25/25 [==============================] - 2s 67ms/step - loss: 0.1339 - acc: 0.9631
Epoch 37/100
25/25 [==============================] - 2s 68ms/step - loss: 0.1160 - acc: 0.9725
Epoch 38/100
25/25 [==============================] - 2s 67ms/step - loss: 0.1098 - acc: 0.9733
Epoch 39/100
25/25 [==============================] - 2s 68ms/step - loss: 0.1240 - acc: 0.9683
Epoch 40/100
25/25 [==============================] - 2s 67ms/step - loss: 0.1179 - acc: 0.9698
Epoch 41/100
25/25 [==============================] - 2s 68ms/step - loss: 0.1110 - acc: 0.9722
Epoch 42/100
25/25 [==============================] - 2s 69ms/step - loss: 0.1091 - acc: 0.9748
Epoch 43/100
25/25 [==============================] - 2s 67ms/step - loss: 0.1032 - acc: 0.9759
Epoch 44/100
25/25 [==============================] - 2s 68ms/step - loss: 0.1042 - acc: 0.9761
Epoch 45/100
25/25 [==============================] - 2s 66ms/step - loss: 0.1050 - acc: 0.9754
Epoch 46/100
25/25 [==============================] -

25/25 [==============================] - 2s 67ms/step - loss: 0.2988 - acc: 0.9012
Epoch 15/100
25/25 [==============================] - 2s 68ms/step - loss: 0.2733 - acc: 0.9148
Epoch 16/100
25/25 [==============================] - 2s 68ms/step - loss: 0.2830 - acc: 0.9044
Epoch 17/100
25/25 [==============================] - 2s 67ms/step - loss: 0.2527 - acc: 0.9228
Epoch 18/100
25/25 [==============================] - 2s 67ms/step - loss: 0.2411 - acc: 0.9266
Epoch 19/100
25/25 [==============================] - 2s 66ms/step - loss: 0.2254 - acc: 0.9325
Epoch 20/100
25/25 [==============================] - 2s 68ms/step - loss: 0.2375 - acc: 0.9270
Epoch 21/100
25/25 [==============================] - 2s 67ms/step - loss: 0.2127 - acc: 0.9395
Epoch 22/100
25/25 [==============================] - 2s 68ms/step - loss: 0.2104 - acc: 0.9408
Epoch 23/100
25/25 [==============================] - 2s 68ms/step - loss: 0.2168 - acc: 0.9366
Epoch 24/100
25/25 [==============================] -

25/25 [==============================] - 2s 66ms/step - loss: 0.1043 - acc: 0.9772
Epoch 98/100
25/25 [==============================] - 2s 66ms/step - loss: 0.1068 - acc: 0.9770
Epoch 99/100
25/25 [==============================] - 2s 66ms/step - loss: 0.1071 - acc: 0.9764
Epoch 100/100
25/25 [==============================] - 2s 66ms/step - loss: 0.1050 - acc: 0.9765
Epoch 1/100
25/25 [==============================] - 5s 68ms/step - loss: 0.6781 - acc: 0.5962
Epoch 2/100
25/25 [==============================] - 2s 68ms/step - loss: 0.6729 - acc: 0.6072
Epoch 3/100
25/25 [==============================] - 2s 69ms/step - loss: 0.6661 - acc: 0.6111
Epoch 4/100
25/25 [==============================] - 2s 68ms/step - loss: 0.5963 - acc: 0.6791
Epoch 5/100
25/25 [==============================] - 2s 68ms/step - loss: 0.5011 - acc: 0.7667
Epoch 6/100
25/25 [==============================] - 2s 68ms/step - loss: 0.4387 - acc: 0.8083
Epoch 7/100
25/25 [==============================] - 2s 67

25/25 [==============================] - 2s 68ms/step - loss: 0.0845 - acc: 0.9816
Epoch 76/100
25/25 [==============================] - 2s 68ms/step - loss: 0.0867 - acc: 0.9798
Epoch 77/100
25/25 [==============================] - 2s 68ms/step - loss: 0.0906 - acc: 0.9785
Epoch 78/100
25/25 [==============================] - 2s 68ms/step - loss: 0.0807 - acc: 0.9825
Epoch 79/100
25/25 [==============================] - 2s 68ms/step - loss: 0.0980 - acc: 0.9733
Epoch 80/100
25/25 [==============================] - 2s 67ms/step - loss: 0.0820 - acc: 0.9820
Epoch 81/100
25/25 [==============================] - 2s 67ms/step - loss: 0.0799 - acc: 0.9827
Epoch 82/100
25/25 [==============================] - 2s 67ms/step - loss: 0.0801 - acc: 0.9814
Epoch 83/100
25/25 [==============================] - 2s 68ms/step - loss: 0.0844 - acc: 0.9796
Epoch 84/100
25/25 [==============================] - 2s 68ms/step - loss: 0.0841 - acc: 0.9808
Epoch 85/100
25/25 [==============================] -

25/25 [==============================] - 2s 67ms/step - loss: 0.1133 - acc: 0.9728
Epoch 53/100
25/25 [==============================] - 2s 66ms/step - loss: 0.1446 - acc: 0.9584
Epoch 54/100
25/25 [==============================] - 2s 67ms/step - loss: 0.1154 - acc: 0.9725
Epoch 55/100
25/25 [==============================] - 2s 68ms/step - loss: 0.1140 - acc: 0.9730
Epoch 56/100
25/25 [==============================] - 2s 67ms/step - loss: 0.1128 - acc: 0.9694
Epoch 57/100
25/25 [==============================] - 2s 67ms/step - loss: 0.1061 - acc: 0.9746
Epoch 58/100
25/25 [==============================] - 2s 68ms/step - loss: 0.0986 - acc: 0.9774
Epoch 59/100
25/25 [==============================] - 2s 68ms/step - loss: 0.1017 - acc: 0.9775
Epoch 60/100
25/25 [==============================] - 2s 69ms/step - loss: 0.1023 - acc: 0.9762
Epoch 61/100
25/25 [==============================] - 2s 67ms/step - loss: 0.0991 - acc: 0.9774
Epoch 62/100
25/25 [==============================] -

25/25 [==============================] - 2s 67ms/step - loss: 0.1366 - acc: 0.9647
Epoch 29/100
25/25 [==============================] - 2s 67ms/step - loss: 0.1269 - acc: 0.9665
Epoch 30/100
25/25 [==============================] - 2s 67ms/step - loss: 0.1235 - acc: 0.9670
Epoch 31/100
25/25 [==============================] - 2s 67ms/step - loss: 0.1192 - acc: 0.9680
Epoch 32/100
25/25 [==============================] - 2s 67ms/step - loss: 0.1281 - acc: 0.9634
Epoch 33/100
25/25 [==============================] - 2s 67ms/step - loss: 0.1357 - acc: 0.9604
Epoch 34/100
25/25 [==============================] - 2s 68ms/step - loss: 0.1137 - acc: 0.9709
Epoch 35/100
25/25 [==============================] - 2s 67ms/step - loss: 0.1048 - acc: 0.9746
Epoch 36/100
25/25 [==============================] - 2s 68ms/step - loss: 0.0987 - acc: 0.9759
Epoch 37/100
25/25 [==============================] - 2s 67ms/step - loss: 0.0998 - acc: 0.9769
Epoch 38/100
25/25 [==============================] -

25/25 [==============================] - 2s 68ms/step - loss: 0.4297 - acc: 0.8135
Epoch 7/100
25/25 [==============================] - 2s 68ms/step - loss: 0.4039 - acc: 0.8321
Epoch 8/100
25/25 [==============================] - 2s 67ms/step - loss: 0.3607 - acc: 0.8562
Epoch 9/100
25/25 [==============================] - 2s 68ms/step - loss: 0.3396 - acc: 0.8699
Epoch 10/100
25/25 [==============================] - 2s 67ms/step - loss: 0.3310 - acc: 0.8748
Epoch 11/100
25/25 [==============================] - 2s 68ms/step - loss: 0.3030 - acc: 0.8913
Epoch 12/100
25/25 [==============================] - 2s 68ms/step - loss: 0.2836 - acc: 0.9002
Epoch 13/100
25/25 [==============================] - 2s 68ms/step - loss: 0.2675 - acc: 0.9097
Epoch 14/100
25/25 [==============================] - 2s 68ms/step - loss: 0.2697 - acc: 0.9076
Epoch 15/100
25/25 [==============================] - 2s 67ms/step - loss: 0.2430 - acc: 0.9225
Epoch 16/100
25/25 [==============================] - 2s

25/25 [==============================] - 2s 67ms/step - loss: 0.0758 - acc: 0.9833
Epoch 91/100
25/25 [==============================] - 2s 67ms/step - loss: 0.0774 - acc: 0.9843
Epoch 92/100
25/25 [==============================] - 2s 67ms/step - loss: 0.0774 - acc: 0.9837
Epoch 93/100
25/25 [==============================] - 2s 67ms/step - loss: 0.0732 - acc: 0.9851
Epoch 94/100
25/25 [==============================] - 2s 68ms/step - loss: 0.0730 - acc: 0.9854
Epoch 95/100
25/25 [==============================] - 2s 67ms/step - loss: 0.0730 - acc: 0.9853
Epoch 96/100
25/25 [==============================] - 2s 67ms/step - loss: 0.0733 - acc: 0.9848
Epoch 97/100
25/25 [==============================] - 2s 66ms/step - loss: 0.0721 - acc: 0.9851
Epoch 98/100
25/25 [==============================] - 2s 66ms/step - loss: 0.0746 - acc: 0.9851
Epoch 99/100
25/25 [==============================] - 2s 67ms/step - loss: 0.0691 - acc: 0.9861
Epoch 100/100
25/25 [==============================] 

25/25 [==============================] - 2s 68ms/step - loss: 0.1130 - acc: 0.9754
Epoch 68/100
25/25 [==============================] - 2s 67ms/step - loss: 0.1118 - acc: 0.9754
Epoch 69/100
25/25 [==============================] - 2s 68ms/step - loss: 0.1115 - acc: 0.9762
Epoch 70/100
25/25 [==============================] - 2s 68ms/step - loss: 0.1139 - acc: 0.9753
Epoch 71/100
25/25 [==============================] - 2s 68ms/step - loss: 0.1203 - acc: 0.9722
Epoch 72/100
25/25 [==============================] - 2s 68ms/step - loss: 0.1133 - acc: 0.9756
Epoch 73/100
25/25 [==============================] - 2s 68ms/step - loss: 0.1117 - acc: 0.9754
Epoch 74/100
25/25 [==============================] - 2s 68ms/step - loss: 0.1166 - acc: 0.9736
Epoch 75/100
25/25 [==============================] - 2s 68ms/step - loss: 0.1173 - acc: 0.9736
Epoch 76/100
25/25 [==============================] - 2s 67ms/step - loss: 0.1380 - acc: 0.9646
Epoch 77/100
25/25 [==============================] -

25/25 [==============================] - 2s 68ms/step - loss: 0.1227 - acc: 0.9712
Epoch 46/100
25/25 [==============================] - 2s 67ms/step - loss: 0.1232 - acc: 0.9712
Epoch 47/100
25/25 [==============================] - 2s 67ms/step - loss: 0.1265 - acc: 0.9701
Epoch 48/100
25/25 [==============================] - 2s 68ms/step - loss: 0.1197 - acc: 0.9720
Epoch 49/100
25/25 [==============================] - 2s 67ms/step - loss: 0.1205 - acc: 0.9719
Epoch 50/100
25/25 [==============================] - 2s 67ms/step - loss: 0.1211 - acc: 0.9728
Epoch 51/100
25/25 [==============================] - 2s 67ms/step - loss: 0.1145 - acc: 0.9749
Epoch 52/100
25/25 [==============================] - 2s 67ms/step - loss: 0.1190 - acc: 0.9707
Epoch 53/100
25/25 [==============================] - 2s 67ms/step - loss: 0.1195 - acc: 0.9728
Epoch 54/100
25/25 [==============================] - 2s 68ms/step - loss: 0.1157 - acc: 0.9736
Epoch 55/100
25/25 [==============================] -

25/25 [==============================] - 2s 67ms/step - loss: 0.1795 - acc: 0.9497
Epoch 24/100
25/25 [==============================] - 2s 67ms/step - loss: 0.1692 - acc: 0.9544
Epoch 25/100
25/25 [==============================] - 2s 66ms/step - loss: 0.1614 - acc: 0.9575
Epoch 26/100
25/25 [==============================] - 2s 67ms/step - loss: 0.1584 - acc: 0.9597
Epoch 27/100
25/25 [==============================] - 2s 68ms/step - loss: 0.1565 - acc: 0.9597
Epoch 28/100
25/25 [==============================] - 2s 67ms/step - loss: 0.1605 - acc: 0.9579
Epoch 29/100
25/25 [==============================] - 2s 67ms/step - loss: 0.1455 - acc: 0.9644
Epoch 30/100
25/25 [==============================] - 2s 66ms/step - loss: 0.1447 - acc: 0.9617
Epoch 31/100
25/25 [==============================] - 2s 68ms/step - loss: 0.1545 - acc: 0.9573
Epoch 32/100
25/25 [==============================] - 2s 67ms/step - loss: 0.1655 - acc: 0.9515
Epoch 33/100
25/25 [==============================] -

Epoch 1/100
25/25 [==============================] - 5s 67ms/step - loss: 0.6921 - acc: 0.5655
Epoch 2/100
25/25 [==============================] - 2s 66ms/step - loss: 0.6736 - acc: 0.6037
Epoch 3/100
25/25 [==============================] - 2s 66ms/step - loss: 0.6706 - acc: 0.6092: 0s - loss: 0.6695 - ac
Epoch 4/100
25/25 [==============================] - 2s 66ms/step - loss: 0.6488 - acc: 0.6170: 0s - loss: 0.6504 - acc: 0.6
Epoch 5/100
25/25 [==============================] - 2s 67ms/step - loss: 0.5634 - acc: 0.7148
Epoch 6/100
25/25 [==============================] - 2s 67ms/step - loss: 0.4687 - acc: 0.7933
Epoch 7/100
25/25 [==============================] - 2s 66ms/step - loss: 0.4351 - acc: 0.8151
Epoch 8/100
25/25 [==============================] - 2s 67ms/step - loss: 0.3887 - acc: 0.8421
Epoch 9/100
25/25 [==============================] - 2s 66ms/step - loss: 0.3665 - acc: 0.8565
Epoch 10/100
25/25 [==============================] - 2s 66ms/step - loss: 0.3400 - acc: 0.

25/25 [==============================] - 2s 68ms/step - loss: 0.0883 - acc: 0.9816
Epoch 85/100
25/25 [==============================] - 2s 66ms/step - loss: 0.0881 - acc: 0.9811
Epoch 86/100
25/25 [==============================] - 2s 68ms/step - loss: 0.0867 - acc: 0.9812
Epoch 87/100
25/25 [==============================] - 2s 67ms/step - loss: 0.0878 - acc: 0.9809
Epoch 88/100
25/25 [==============================] - 2s 67ms/step - loss: 0.0855 - acc: 0.9816
Epoch 89/100
25/25 [==============================] - 2s 67ms/step - loss: 0.0872 - acc: 0.9816
Epoch 90/100
25/25 [==============================] - 2s 69ms/step - loss: 0.0871 - acc: 0.9811
Epoch 91/100
25/25 [==============================] - 2s 68ms/step - loss: 0.0913 - acc: 0.9804
Epoch 92/100
25/25 [==============================] - 2s 68ms/step - loss: 0.0855 - acc: 0.9824
Epoch 93/100
25/25 [==============================] - 2s 67ms/step - loss: 0.0840 - acc: 0.9822
Epoch 94/100
25/25 [==============================] -

25/25 [==============================] - 2s 67ms/step - loss: 0.0913 - acc: 0.9775: 1s - loss: 
Epoch 63/100
25/25 [==============================] - 2s 68ms/step - loss: 0.0905 - acc: 0.9791
Epoch 64/100
25/25 [==============================] - 2s 68ms/step - loss: 0.1014 - acc: 0.9741
Epoch 65/100
25/25 [==============================] - 2s 68ms/step - loss: 0.0896 - acc: 0.9798
Epoch 66/100
25/25 [==============================] - 2s 68ms/step - loss: 0.0859 - acc: 0.9804
Epoch 67/100
25/25 [==============================] - 2s 68ms/step - loss: 0.0844 - acc: 0.9812
Epoch 68/100
25/25 [==============================] - 2s 67ms/step - loss: 0.0831 - acc: 0.9816
Epoch 69/100
25/25 [==============================] - 2s 67ms/step - loss: 0.0790 - acc: 0.9833: 0s - loss: 0.0785 - ac
Epoch 70/100
25/25 [==============================] - 2s 67ms/step - loss: 0.0799 - acc: 0.9825
Epoch 71/100
25/25 [==============================] - 2s 67ms/step - loss: 0.0828 - acc: 0.9817
Epoch 72/100
25/

Epoch 39/100
25/25 [==============================] - 2s 67ms/step - loss: 0.1256 - acc: 0.9693
Epoch 40/100
25/25 [==============================] - 2s 66ms/step - loss: 0.1223 - acc: 0.9704
Epoch 41/100
25/25 [==============================] - 2s 68ms/step - loss: 0.1166 - acc: 0.9715
Epoch 42/100
25/25 [==============================] - 2s 67ms/step - loss: 0.1155 - acc: 0.9731
Epoch 43/100
25/25 [==============================] - 2s 67ms/step - loss: 0.1136 - acc: 0.9740
Epoch 44/100
25/25 [==============================] - 2s 67ms/step - loss: 0.1111 - acc: 0.9744
Epoch 45/100
25/25 [==============================] - 2s 65ms/step - loss: 0.1089 - acc: 0.9749
Epoch 46/100
25/25 [==============================] - 2s 67ms/step - loss: 0.1083 - acc: 0.9756
Epoch 47/100
25/25 [==============================] - 2s 67ms/step - loss: 0.1088 - acc: 0.9743
Epoch 48/100
25/25 [==============================] - 2s 67ms/step - loss: 0.1129 - acc: 0.9731
Epoch 49/100
25/25 [====================

25/25 [==============================] - 2s 66ms/step - loss: 0.2516 - acc: 0.9253
Epoch 18/100
25/25 [==============================] - 2s 68ms/step - loss: 0.2436 - acc: 0.9235
Epoch 19/100
25/25 [==============================] - 2s 68ms/step - loss: 0.2279 - acc: 0.9321
Epoch 20/100
25/25 [==============================] - 2s 66ms/step - loss: 0.2358 - acc: 0.9227
Epoch 21/100
25/25 [==============================] - 2s 67ms/step - loss: 0.2128 - acc: 0.9389
Epoch 22/100
25/25 [==============================] - 2s 67ms/step - loss: 0.2075 - acc: 0.9405
Epoch 23/100
25/25 [==============================] - 2s 67ms/step - loss: 0.2015 - acc: 0.9421
Epoch 24/100
25/25 [==============================] - 2s 67ms/step - loss: 0.1959 - acc: 0.9445
Epoch 25/100
25/25 [==============================] - 2s 68ms/step - loss: 0.1836 - acc: 0.9513
Epoch 26/100
25/25 [==============================] - 2s 67ms/step - loss: 0.1802 - acc: 0.9524: 0s - loss: 0.1808 - acc: 0.
Epoch 27/100
25/25 [====

Epoch 1/100
25/25 [==============================] - 4s 67ms/step - loss: 0.6732 - acc: 0.6043
Epoch 2/100
25/25 [==============================] - 2s 66ms/step - loss: 0.6710 - acc: 0.6061
Epoch 3/100
25/25 [==============================] - 2s 68ms/step - loss: 0.6601 - acc: 0.6127
Epoch 4/100
25/25 [==============================] - 2s 67ms/step - loss: 0.5781 - acc: 0.6995
Epoch 5/100
25/25 [==============================] - 2s 66ms/step - loss: 0.4854 - acc: 0.7790
Epoch 6/100
25/25 [==============================] - 2s 67ms/step - loss: 0.4289 - acc: 0.8120
Epoch 7/100
25/25 [==============================] - 2s 66ms/step - loss: 0.4060 - acc: 0.8303
Epoch 8/100
25/25 [==============================] - 2s 67ms/step - loss: 0.3638 - acc: 0.8570
Epoch 9/100
25/25 [==============================] - 2s 66ms/step - loss: 0.3406 - acc: 0.8678
Epoch 10/100
25/25 [==============================] - 2s 67ms/step - loss: 0.3203 - acc: 0.8797
Epoch 11/100
25/25 [=============================

25/25 [==============================] - 2s 68ms/step - loss: 0.0755 - acc: 0.9840
Epoch 81/100
25/25 [==============================] - 2s 68ms/step - loss: 0.0743 - acc: 0.9846
Epoch 82/100
25/25 [==============================] - 2s 67ms/step - loss: 0.0711 - acc: 0.9854
Epoch 83/100
25/25 [==============================] - 2s 68ms/step - loss: 0.0715 - acc: 0.9846
Epoch 84/100
25/25 [==============================] - 2s 68ms/step - loss: 0.0740 - acc: 0.9848
Epoch 85/100
25/25 [==============================] - 2s 68ms/step - loss: 0.0737 - acc: 0.9848
Epoch 86/100
25/25 [==============================] - 2s 67ms/step - loss: 0.0708 - acc: 0.9854
Epoch 87/100
25/25 [==============================] - 2s 67ms/step - loss: 0.0703 - acc: 0.9859
Epoch 88/100
25/25 [==============================] - 2s 67ms/step - loss: 0.0663 - acc: 0.9871
Epoch 89/100
25/25 [==============================] - 2s 67ms/step - loss: 0.0647 - acc: 0.9871
Epoch 90/100
25/25 [==============================] -

25/25 [==============================] - 2s 67ms/step - loss: 0.1209 - acc: 0.9728
Epoch 59/100
25/25 [==============================] - 2s 68ms/step - loss: 0.1225 - acc: 0.9715
Epoch 60/100
25/25 [==============================] - 2s 67ms/step - loss: 0.1240 - acc: 0.9706
Epoch 61/100
25/25 [==============================] - 2s 68ms/step - loss: 0.1181 - acc: 0.9735
Epoch 62/100
25/25 [==============================] - 2s 67ms/step - loss: 0.1211 - acc: 0.9712
Epoch 63/100
25/25 [==============================] - 2s 68ms/step - loss: 0.1281 - acc: 0.9688
Epoch 64/100
25/25 [==============================] - 2s 67ms/step - loss: 0.1234 - acc: 0.9722
Epoch 65/100
25/25 [==============================] - 2s 67ms/step - loss: 0.1190 - acc: 0.9733
Epoch 66/100
25/25 [==============================] - 2s 67ms/step - loss: 0.1184 - acc: 0.9731
Epoch 67/100
25/25 [==============================] - 2s 67ms/step - loss: 0.1176 - acc: 0.9735
Epoch 68/100
25/25 [==============================] -

25/25 [==============================] - 2s 67ms/step - loss: 0.1216 - acc: 0.9688
Epoch 35/100
25/25 [==============================] - 2s 66ms/step - loss: 0.1164 - acc: 0.9722
Epoch 36/100
25/25 [==============================] - 2s 68ms/step - loss: 0.1130 - acc: 0.9740
Epoch 37/100
25/25 [==============================] - 2s 67ms/step - loss: 0.1124 - acc: 0.9736
Epoch 38/100
25/25 [==============================] - 2s 68ms/step - loss: 0.1066 - acc: 0.9762
Epoch 39/100
25/25 [==============================] - 2s 67ms/step - loss: 0.1146 - acc: 0.9723
Epoch 40/100
25/25 [==============================] - 2s 67ms/step - loss: 0.1162 - acc: 0.9699
Epoch 41/100
25/25 [==============================] - 2s 67ms/step - loss: 0.1065 - acc: 0.9751: 0s - loss: 0.1218 
Epoch 42/100
25/25 [==============================] - 2s 67ms/step - loss: 0.1023 - acc: 0.9767
Epoch 43/100
25/25 [==============================] - 2s 68ms/step - loss: 0.1007 - acc: 0.9774
Epoch 44/100
25/25 [=============

25/25 [==============================] - 2s 66ms/step - loss: 0.3263 - acc: 0.8818
Epoch 11/100
25/25 [==============================] - 2s 67ms/step - loss: 0.3089 - acc: 0.8892
Epoch 12/100
25/25 [==============================] - 2s 67ms/step - loss: 0.2850 - acc: 0.8984
Epoch 13/100
25/25 [==============================] - 2s 67ms/step - loss: 0.2680 - acc: 0.9112
Epoch 14/100
25/25 [==============================] - 2s 67ms/step - loss: 0.2810 - acc: 0.8999
Epoch 15/100
25/25 [==============================] - 2s 66ms/step - loss: 0.2530 - acc: 0.9156
Epoch 16/100
25/25 [==============================] - 2s 68ms/step - loss: 0.2345 - acc: 0.9249
Epoch 17/100
25/25 [==============================] - 2s 67ms/step - loss: 0.2164 - acc: 0.9343
Epoch 18/100
25/25 [==============================] - 2s 67ms/step - loss: 0.2224 - acc: 0.9296
Epoch 19/100
25/25 [==============================] - 2s 66ms/step - loss: 0.2044 - acc: 0.9387
Epoch 20/100
25/25 [==============================] -

25/25 [==============================] - 2s 67ms/step - loss: 0.0847 - acc: 0.9811
Epoch 96/100
25/25 [==============================] - 2s 66ms/step - loss: 0.0807 - acc: 0.9833
Epoch 97/100
25/25 [==============================] - 2s 66ms/step - loss: 0.0830 - acc: 0.9825
Epoch 98/100
25/25 [==============================] - 2s 66ms/step - loss: 0.0847 - acc: 0.9811: 1s - loss
Epoch 99/100
25/25 [==============================] - 2s 66ms/step - loss: 0.0798 - acc: 0.9833
Epoch 100/100
25/25 [==============================] - 2s 66ms/step - loss: 0.0801 - acc: 0.9829
Epoch 1/100
25/25 [==============================] - 5s 66ms/step - loss: 0.7436 - acc: 0.5110
Epoch 2/100
25/25 [==============================] - 2s 67ms/step - loss: 0.6773 - acc: 0.6040: 1s - loss: 0.6752 - - ETA: 0s - loss: 0.6766 - acc: 0.61 - ETA: 0s - loss: 0.6782 - acc:
Epoch 3/100
25/25 [==============================] - 2s 67ms/step - loss: 0.6755 - acc: 0.5979
Epoch 4/100
25/25 [==============================]

25/25 [==============================] - 2s 66ms/step - loss: 0.1167 - acc: 0.9730
Epoch 71/100
25/25 [==============================] - 2s 68ms/step - loss: 0.1183 - acc: 0.9722
Epoch 72/100
25/25 [==============================] - 2s 67ms/step - loss: 0.1159 - acc: 0.9740
Epoch 73/100
25/25 [==============================] - 2s 67ms/step - loss: 0.1099 - acc: 0.9749
Epoch 74/100
25/25 [==============================] - 2s 68ms/step - loss: 0.1104 - acc: 0.9753
Epoch 75/100
25/25 [==============================] - 2s 67ms/step - loss: 0.1107 - acc: 0.9762
Epoch 76/100
25/25 [==============================] - 2s 68ms/step - loss: 0.1189 - acc: 0.9715
Epoch 77/100
25/25 [==============================] - 2s 67ms/step - loss: 0.1113 - acc: 0.9754
Epoch 78/100
25/25 [==============================] - 2s 68ms/step - loss: 0.1123 - acc: 0.9751
Epoch 79/100
25/25 [==============================] - 2s 67ms/step - loss: 0.1080 - acc: 0.9774
Epoch 80/100
25/25 [==============================] -

25/25 [==============================] - 2s 68ms/step - loss: 0.1256 - acc: 0.9694
Epoch 49/100
25/25 [==============================] - 2s 67ms/step - loss: 0.1326 - acc: 0.9683
Epoch 50/100
25/25 [==============================] - 2s 67ms/step - loss: 0.1258 - acc: 0.9710
Epoch 51/100
25/25 [==============================] - 2s 67ms/step - loss: 0.1283 - acc: 0.9699: 0s - loss: 0.1283 - acc: 0.969
Epoch 52/100
25/25 [==============================] - 2s 68ms/step - loss: 0.1219 - acc: 0.9715
Epoch 53/100
25/25 [==============================] - 2s 67ms/step - loss: 0.1231 - acc: 0.9715
Epoch 54/100
25/25 [==============================] - 2s 67ms/step - loss: 0.1183 - acc: 0.9723
Epoch 55/100
25/25 [==============================] - 2s 68ms/step - loss: 0.1180 - acc: 0.9720
Epoch 56/100
25/25 [==============================] - 2s 69ms/step - loss: 0.1185 - acc: 0.9723
Epoch 57/100
25/25 [==============================] - 2s 67ms/step - loss: 0.1188 - acc: 0.9728
Epoch 58/100
25/25 [=

25/25 [==============================] - 2s 67ms/step - loss: 0.1939 - acc: 0.9432
Epoch 24/100
25/25 [==============================] - 2s 68ms/step - loss: 0.1746 - acc: 0.9528
Epoch 25/100
25/25 [==============================] - 2s 67ms/step - loss: 0.1678 - acc: 0.9566
Epoch 26/100
25/25 [==============================] - 2s 67ms/step - loss: 0.1706 - acc: 0.9549
Epoch 27/100
25/25 [==============================] - 2s 67ms/step - loss: 0.1682 - acc: 0.9563
Epoch 28/100
25/25 [==============================] - 2s 68ms/step - loss: 0.1621 - acc: 0.9583
Epoch 29/100
25/25 [==============================] - 2s 67ms/step - loss: 0.1628 - acc: 0.9555
Epoch 30/100
25/25 [==============================] - 2s 67ms/step - loss: 0.1578 - acc: 0.9597: 1s - loss: 0.1740 - acc: 0. - ETA: 1s - loss: 0
Epoch 31/100
25/25 [==============================] - 2s 67ms/step - loss: 0.1536 - acc: 0.9610
Epoch 32/100
25/25 [==============================] - 2s 67ms/step - loss: 0.1637 - acc: 0.9545
Epoc

25/25 [==============================] - 5s 66ms/step - loss: 0.6736 - acc: 0.6081
Epoch 2/100
25/25 [==============================] - 2s 67ms/step - loss: 0.6709 - acc: 0.6082
Epoch 3/100
25/25 [==============================] - 2s 66ms/step - loss: 0.6627 - acc: 0.6093
Epoch 4/100
25/25 [==============================] - 2s 67ms/step - loss: 0.5874 - acc: 0.6886
Epoch 5/100
25/25 [==============================] - 2s 66ms/step - loss: 0.4990 - acc: 0.7724
Epoch 6/100
25/25 [==============================] - 2s 67ms/step - loss: 0.4367 - acc: 0.8101
Epoch 7/100
25/25 [==============================] - 2s 67ms/step - loss: 0.4071 - acc: 0.8303
Epoch 8/100
25/25 [==============================] - 2s 67ms/step - loss: 0.3655 - acc: 0.8549
Epoch 9/100
25/25 [==============================] - 2s 66ms/step - loss: 0.3422 - acc: 0.8682: 1s - loss: 0.3254 - acc: - ETA: 0s - loss: 0.326
Epoch 10/100
25/25 [==============================] - 2s 67ms/step - loss: 0.3254 - acc: 0.8818
Epoch 11/10

25/25 [==============================] - 2s 67ms/step - loss: 0.0886 - acc: 0.9822: 1s - loss: 0.0887 - acc: 0.981 - ETA: 1s - loss: 0
Epoch 86/100
25/25 [==============================] - 2s 67ms/step - loss: 0.0904 - acc: 0.9804
Epoch 87/100
25/25 [==============================] - 2s 67ms/step - loss: 0.0895 - acc: 0.9803
Epoch 88/100
25/25 [==============================] - 2s 68ms/step - loss: 0.0876 - acc: 0.9817
Epoch 89/100
25/25 [==============================] - 2s 67ms/step - loss: 0.0863 - acc: 0.9819
Epoch 90/100
25/25 [==============================] - 2s 67ms/step - loss: 0.0826 - acc: 0.9822
Epoch 91/100
25/25 [==============================] - 2s 67ms/step - loss: 0.0897 - acc: 0.9806
Epoch 92/100
25/25 [==============================] - 2s 67ms/step - loss: 0.0888 - acc: 0.9809
Epoch 93/100
25/25 [==============================] - 2s 67ms/step - loss: 0.0865 - acc: 0.9822
Epoch 94/100
25/25 [==============================] - 2s 67ms/step - loss: 0.0860 - acc: 0.9814
E

25/25 [==============================] - 2s 70ms/step - loss: 0.1191 - acc: 0.9704
Epoch 64/100
25/25 [==============================] - 2s 71ms/step - loss: 0.1144 - acc: 0.9720
Epoch 65/100
25/25 [==============================] - 2s 70ms/step - loss: 0.1048 - acc: 0.9770
Epoch 66/100
25/25 [==============================] - 2s 69ms/step - loss: 0.1007 - acc: 0.9769
Epoch 67/100
25/25 [==============================] - 2s 71ms/step - loss: 0.0987 - acc: 0.9782
Epoch 68/100
25/25 [==============================] - 2s 70ms/step - loss: 0.0973 - acc: 0.9782
Epoch 69/100
25/25 [==============================] - 2s 70ms/step - loss: 0.0967 - acc: 0.9783
Epoch 70/100
25/25 [==============================] - 2s 71ms/step - loss: 0.1006 - acc: 0.9780
Epoch 71/100
25/25 [==============================] - 2s 70ms/step - loss: 0.1047 - acc: 0.9756
Epoch 72/100
25/25 [==============================] - 2s 70ms/step - loss: 0.0987 - acc: 0.9780
Epoch 73/100
25/25 [==============================] -

25/25 [==============================] - 2s 70ms/step - loss: 0.1270 - acc: 0.9683
Epoch 42/100
25/25 [==============================] - 2s 71ms/step - loss: 0.1402 - acc: 0.9609
Epoch 43/100
25/25 [==============================] - 2s 69ms/step - loss: 0.1236 - acc: 0.9698
Epoch 44/100
25/25 [==============================] - 2s 70ms/step - loss: 0.1177 - acc: 0.9715
Epoch 45/100
25/25 [==============================] - 2s 68ms/step - loss: 0.1180 - acc: 0.9701
Epoch 46/100
25/25 [==============================] - 2s 70ms/step - loss: 0.1102 - acc: 0.9751
Epoch 47/100
25/25 [==============================] - 2s 70ms/step - loss: 0.1086 - acc: 0.9759
Epoch 48/100
25/25 [==============================] - 2s 69ms/step - loss: 0.1086 - acc: 0.9740
Epoch 49/100
25/25 [==============================] - 2s 70ms/step - loss: 0.1083 - acc: 0.9749
Epoch 50/100
25/25 [==============================] - 2s 71ms/step - loss: 0.1072 - acc: 0.9757
Epoch 51/100
25/25 [==============================] -

25/25 [==============================] - 2s 67ms/step - loss: 0.2052 - acc: 0.9380
Epoch 20/100
25/25 [==============================] - 2s 73ms/step - loss: 0.1933 - acc: 0.9453
Epoch 21/100
25/25 [==============================] - 2s 70ms/step - loss: 0.1848 - acc: 0.9474
Epoch 22/100
25/25 [==============================] - 2s 71ms/step - loss: 0.1834 - acc: 0.9471
Epoch 23/100
25/25 [==============================] - 2s 69ms/step - loss: 0.1715 - acc: 0.9537
Epoch 24/100
25/25 [==============================] - 2s 70ms/step - loss: 0.1693 - acc: 0.9528
Epoch 25/100
25/25 [==============================] - 2s 71ms/step - loss: 0.1749 - acc: 0.9487
Epoch 26/100
25/25 [==============================] - 2s 69ms/step - loss: 0.1688 - acc: 0.9502
Epoch 27/100
25/25 [==============================] - 2s 69ms/step - loss: 0.1550 - acc: 0.9578
Epoch 28/100
25/25 [==============================] - 2s 79ms/step - loss: 0.1519 - acc: 0.9605
Epoch 29/100
25/25 [==============================] -

Epoch 1/100
25/25 [==============================] - 5s 73ms/step - loss: 0.6935 - acc: 0.5657
Epoch 2/100
25/25 [==============================] - 2s 73ms/step - loss: 0.6749 - acc: 0.5992
Epoch 3/100
25/25 [==============================] - 2s 74ms/step - loss: 0.6709 - acc: 0.6116
Epoch 4/100
25/25 [==============================] - 2s 72ms/step - loss: 0.6448 - acc: 0.6205
Epoch 5/100
25/25 [==============================] - 2s 72ms/step - loss: 0.5549 - acc: 0.7214
Epoch 6/100
25/25 [==============================] - 2s 72ms/step - loss: 0.4623 - acc: 0.7965
Epoch 7/100
25/25 [==============================] - 2s 73ms/step - loss: 0.4348 - acc: 0.8133
Epoch 8/100
25/25 [==============================] - 2s 71ms/step - loss: 0.3927 - acc: 0.8381
Epoch 9/100
25/25 [==============================] - 2s 72ms/step - loss: 0.3670 - acc: 0.8565
Epoch 10/100
25/25 [==============================] - 2s 71ms/step - loss: 0.3373 - acc: 0.8733
Epoch 11/100
25/25 [=============================

25/25 [==============================] - 2s 73ms/step - loss: 0.0988 - acc: 0.9772
Epoch 85/100
25/25 [==============================] - 2s 73ms/step - loss: 0.0949 - acc: 0.9798
Epoch 86/100
25/25 [==============================] - 2s 73ms/step - loss: 0.0959 - acc: 0.9788
Epoch 87/100
25/25 [==============================] - 2s 73ms/step - loss: 0.0921 - acc: 0.9812
Epoch 88/100
25/25 [==============================] - 2s 74ms/step - loss: 0.0900 - acc: 0.9816
Epoch 89/100
25/25 [==============================] - 2s 73ms/step - loss: 0.0919 - acc: 0.9801
Epoch 90/100
25/25 [==============================] - 2s 73ms/step - loss: 0.0884 - acc: 0.9816
Epoch 91/100
25/25 [==============================] - 2s 74ms/step - loss: 0.0914 - acc: 0.9809
Epoch 92/100
25/25 [==============================] - 2s 74ms/step - loss: 0.0919 - acc: 0.9791
Epoch 93/100
25/25 [==============================] - 2s 72ms/step - loss: 0.0908 - acc: 0.9804
Epoch 94/100
25/25 [==============================] -

25/25 [==============================] - 2s 73ms/step - loss: 0.0952 - acc: 0.9777
Epoch 63/100
25/25 [==============================] - 2s 76ms/step - loss: 0.1036 - acc: 0.9753
Epoch 64/100
25/25 [==============================] - 2s 73ms/step - loss: 0.0936 - acc: 0.9788
Epoch 65/100
25/25 [==============================] - 2s 72ms/step - loss: 0.0966 - acc: 0.9772
Epoch 66/100
25/25 [==============================] - 2s 74ms/step - loss: 0.0929 - acc: 0.9799
Epoch 67/100
25/25 [==============================] - 2s 74ms/step - loss: 0.0923 - acc: 0.9793
Epoch 68/100
25/25 [==============================] - 2s 74ms/step - loss: 0.0882 - acc: 0.9798
Epoch 69/100
25/25 [==============================] - 2s 73ms/step - loss: 0.0863 - acc: 0.9814
Epoch 70/100
25/25 [==============================] - 2s 74ms/step - loss: 0.0873 - acc: 0.9809
Epoch 71/100
25/25 [==============================] - 2s 73ms/step - loss: 0.0872 - acc: 0.9801
Epoch 72/100
25/25 [==============================] -

25/25 [==============================] - 2s 74ms/step - loss: 0.1133 - acc: 0.9719
Epoch 41/100
25/25 [==============================] - 2s 73ms/step - loss: 0.1261 - acc: 0.9646
Epoch 42/100
25/25 [==============================] - 2s 74ms/step - loss: 0.1155 - acc: 0.9701
Epoch 43/100
25/25 [==============================] - 2s 74ms/step - loss: 0.1053 - acc: 0.9748
Epoch 44/100
25/25 [==============================] - 2s 75ms/step - loss: 0.1059 - acc: 0.9751
Epoch 45/100
25/25 [==============================] - 2s 74ms/step - loss: 0.1068 - acc: 0.9757
Epoch 46/100
25/25 [==============================] - 2s 74ms/step - loss: 0.1146 - acc: 0.9706
Epoch 47/100
25/25 [==============================] - 2s 73ms/step - loss: 0.1014 - acc: 0.9759
Epoch 48/100
25/25 [==============================] - 2s 74ms/step - loss: 0.1136 - acc: 0.9693
Epoch 49/100
25/25 [==============================] - 2s 72ms/step - loss: 0.1017 - acc: 0.9769
Epoch 50/100
25/25 [==============================] -

25/25 [==============================] - 2s 72ms/step - loss: 0.2136 - acc: 0.9258
Epoch 19/100
25/25 [==============================] - 2s 73ms/step - loss: 0.1875 - acc: 0.9413
Epoch 20/100
25/25 [==============================] - 2s 73ms/step - loss: 0.1747 - acc: 0.9482
Epoch 21/100
25/25 [==============================] - 2s 73ms/step - loss: 0.1760 - acc: 0.9448
Epoch 22/100
25/25 [==============================] - 2s 71ms/step - loss: 0.1637 - acc: 0.9510
Epoch 23/100
25/25 [==============================] - 2s 72ms/step - loss: 0.1568 - acc: 0.9510
Epoch 24/100
25/25 [==============================] - 2s 80ms/step - loss: 0.1461 - acc: 0.9573
Epoch 25/100
25/25 [==============================] - 2s 74ms/step - loss: 0.1307 - acc: 0.9654
Epoch 26/100
25/25 [==============================] - 2s 74ms/step - loss: 0.1282 - acc: 0.9655
Epoch 27/100
25/25 [==============================] - 2s 74ms/step - loss: 0.1185 - acc: 0.9683
Epoch 28/100
25/25 [==============================] -

Epoch 1/100
25/25 [==============================] - 5s 72ms/step - loss: 0.6731 - acc: 0.6090
Epoch 2/100
25/25 [==============================] - 2s 74ms/step - loss: 0.6708 - acc: 0.6098
Epoch 3/100
25/25 [==============================] - 2s 72ms/step - loss: 0.6609 - acc: 0.6126
Epoch 4/100
25/25 [==============================] - 2s 72ms/step - loss: 0.5852 - acc: 0.6964
Epoch 5/100
25/25 [==============================] - 2s 72ms/step - loss: 0.4973 - acc: 0.7705
Epoch 6/100
25/25 [==============================] - 2s 73ms/step - loss: 0.4401 - acc: 0.8141
Epoch 7/100
25/25 [==============================] - 2s 74ms/step - loss: 0.4029 - acc: 0.8303
Epoch 8/100
25/25 [==============================] - 2s 74ms/step - loss: 0.3651 - acc: 0.8528
Epoch 9/100
25/25 [==============================] - 2s 75ms/step - loss: 0.3447 - acc: 0.8724
Epoch 10/100
25/25 [==============================] - 2s 73ms/step - loss: 0.3278 - acc: 0.8787
Epoch 11/100
25/25 [=============================

25/25 [==============================] - 2s 73ms/step - loss: 0.0984 - acc: 0.9782
Epoch 84/100
25/25 [==============================] - 2s 74ms/step - loss: 0.0949 - acc: 0.9798
Epoch 85/100
25/25 [==============================] - 2s 73ms/step - loss: 0.0986 - acc: 0.9775
Epoch 86/100
25/25 [==============================] - 2s 74ms/step - loss: 0.1018 - acc: 0.9759
Epoch 87/100
25/25 [==============================] - 2s 74ms/step - loss: 0.0950 - acc: 0.9786
Epoch 88/100
25/25 [==============================] - 2s 74ms/step - loss: 0.0936 - acc: 0.9803
Epoch 89/100
25/25 [==============================] - 2s 74ms/step - loss: 0.0915 - acc: 0.9806
Epoch 90/100
25/25 [==============================] - 2s 72ms/step - loss: 0.0919 - acc: 0.9801
Epoch 91/100
25/25 [==============================] - 2s 74ms/step - loss: 0.0912 - acc: 0.9811
Epoch 92/100
25/25 [==============================] - 2s 73ms/step - loss: 0.0942 - acc: 0.9801
Epoch 93/100
25/25 [==============================] -

25/25 [==============================] - 2s 71ms/step - loss: 0.0816 - acc: 0.9816
Epoch 62/100
25/25 [==============================] - 2s 71ms/step - loss: 0.0817 - acc: 0.9803
Epoch 63/100
25/25 [==============================] - 2s 72ms/step - loss: 0.0880 - acc: 0.9791
Epoch 64/100
25/25 [==============================] - 2s 72ms/step - loss: 0.0843 - acc: 0.9803
Epoch 65/100
25/25 [==============================] - 2s 71ms/step - loss: 0.0855 - acc: 0.9798
Epoch 66/100
25/25 [==============================] - 2s 72ms/step - loss: 0.0866 - acc: 0.9796
Epoch 67/100
25/25 [==============================] - 2s 70ms/step - loss: 0.0813 - acc: 0.9811
Epoch 68/100
25/25 [==============================] - 2s 71ms/step - loss: 0.0788 - acc: 0.9812
Epoch 69/100
25/25 [==============================] - 2s 71ms/step - loss: 0.0770 - acc: 0.9833
Epoch 70/100
25/25 [==============================] - 2s 72ms/step - loss: 0.0814 - acc: 0.9816
Epoch 71/100
25/25 [==============================] -

25/25 [==============================] - 2s 70ms/step - loss: 0.0942 - acc: 0.9761
Epoch 39/100
25/25 [==============================] - 2s 70ms/step - loss: 0.0967 - acc: 0.9751
Epoch 40/100
25/25 [==============================] - 2s 69ms/step - loss: 0.0906 - acc: 0.9761
Epoch 41/100
25/25 [==============================] - 2s 71ms/step - loss: 0.0939 - acc: 0.9754
Epoch 42/100
25/25 [==============================] - 2s 69ms/step - loss: 0.0840 - acc: 0.9786
Epoch 43/100
25/25 [==============================] - 2s 70ms/step - loss: 0.0828 - acc: 0.9811
Epoch 44/100
25/25 [==============================] - 2s 71ms/step - loss: 0.0815 - acc: 0.9811
Epoch 45/100
25/25 [==============================] - 2s 72ms/step - loss: 0.0851 - acc: 0.9785
Epoch 46/100
25/25 [==============================] - 2s 72ms/step - loss: 0.0797 - acc: 0.9795
Epoch 47/100
25/25 [==============================] - 2s 73ms/step - loss: 0.0815 - acc: 0.9791
Epoch 48/100
25/25 [==============================] -

25/25 [==============================] - 2s 74ms/step - loss: 0.2408 - acc: 0.9165
Epoch 17/100
25/25 [==============================] - 2s 75ms/step - loss: 0.2112 - acc: 0.9334
Epoch 18/100
25/25 [==============================] - 2s 73ms/step - loss: 0.2028 - acc: 0.9356
Epoch 19/100
25/25 [==============================] - 2s 73ms/step - loss: 0.1991 - acc: 0.9359
Epoch 20/100
25/25 [==============================] - 2s 74ms/step - loss: 0.1774 - acc: 0.9463
Epoch 21/100
25/25 [==============================] - 2s 72ms/step - loss: 0.1774 - acc: 0.9474
Epoch 22/100
25/25 [==============================] - 2s 73ms/step - loss: 0.1661 - acc: 0.9524
Epoch 23/100
25/25 [==============================] - 2s 73ms/step - loss: 0.1622 - acc: 0.9529
Epoch 24/100
25/25 [==============================] - 2s 73ms/step - loss: 0.1474 - acc: 0.9591
Epoch 25/100
25/25 [==============================] - 2s 73ms/step - loss: 0.1426 - acc: 0.9599
Epoch 26/100
25/25 [==============================] -

Epoch 1/100
25/25 [==============================] - 5s 73ms/step - loss: 0.6784 - acc: 0.5984
Epoch 2/100
25/25 [==============================] - 2s 73ms/step - loss: 0.6735 - acc: 0.6068
Epoch 3/100
25/25 [==============================] - 2s 72ms/step - loss: 0.6563 - acc: 0.6182
Epoch 4/100
25/25 [==============================] - 2s 72ms/step - loss: 0.5683 - acc: 0.7143
Epoch 5/100
25/25 [==============================] - 2s 72ms/step - loss: 0.4840 - acc: 0.7789
Epoch 6/100
25/25 [==============================] - 2s 74ms/step - loss: 0.4303 - acc: 0.8120
Epoch 7/100
25/25 [==============================] - 2s 74ms/step - loss: 0.3938 - acc: 0.8357
Epoch 8/100
25/25 [==============================] - 2s 74ms/step - loss: 0.3634 - acc: 0.8552
Epoch 9/100
25/25 [==============================] - 2s 73ms/step - loss: 0.3353 - acc: 0.8719
Epoch 10/100
25/25 [==============================] - 2s 74ms/step - loss: 0.3147 - acc: 0.8821
Epoch 11/100
25/25 [=============================

25/25 [==============================] - 2s 73ms/step - loss: 0.0729 - acc: 0.9832
Epoch 82/100
25/25 [==============================] - 2s 73ms/step - loss: 0.0870 - acc: 0.9769
Epoch 83/100
25/25 [==============================] - 2s 73ms/step - loss: 0.0674 - acc: 0.9858
Epoch 84/100
25/25 [==============================] - 2s 73ms/step - loss: 0.0665 - acc: 0.9863
Epoch 85/100
25/25 [==============================] - 2s 74ms/step - loss: 0.0652 - acc: 0.9858
Epoch 86/100
25/25 [==============================] - 2s 74ms/step - loss: 0.0682 - acc: 0.9856
Epoch 87/100
25/25 [==============================] - 2s 75ms/step - loss: 0.0716 - acc: 0.9825
Epoch 88/100
25/25 [==============================] - 2s 73ms/step - loss: 0.0711 - acc: 0.9833
Epoch 89/100
25/25 [==============================] - 2s 72ms/step - loss: 0.0658 - acc: 0.9861
Epoch 90/100
25/25 [==============================] - 2s 73ms/step - loss: 0.0643 - acc: 0.9871
Epoch 91/100
25/25 [==============================] -

25/25 [==============================] - 2s 73ms/step - loss: 0.1044 - acc: 0.9746
Epoch 60/100
25/25 [==============================] - 2s 74ms/step - loss: 0.1025 - acc: 0.9759
Epoch 61/100
25/25 [==============================] - 2s 73ms/step - loss: 0.0919 - acc: 0.9793
Epoch 62/100
25/25 [==============================] - 2s 73ms/step - loss: 0.0946 - acc: 0.9778
Epoch 63/100
25/25 [==============================] - 2s 71ms/step - loss: 0.0952 - acc: 0.9778
Epoch 64/100
25/25 [==============================] - 2s 73ms/step - loss: 0.0890 - acc: 0.9804
Epoch 65/100
25/25 [==============================] - 2s 74ms/step - loss: 0.0889 - acc: 0.9811
Epoch 66/100
25/25 [==============================] - 2s 73ms/step - loss: 0.0908 - acc: 0.9795
Epoch 67/100
25/25 [==============================] - 2s 73ms/step - loss: 0.0871 - acc: 0.9809
Epoch 68/100
25/25 [==============================] - 2s 74ms/step - loss: 0.0868 - acc: 0.9816
Epoch 69/100
25/25 [==============================] -

25/25 [==============================] - 2s 68ms/step - loss: 0.1167 - acc: 0.9728
Epoch 38/100
25/25 [==============================] - 2s 78ms/step - loss: 0.1118 - acc: 0.9741
Epoch 39/100
25/25 [==============================] - 2s 70ms/step - loss: 0.1224 - acc: 0.9676
Epoch 40/100
25/25 [==============================] - 2s 71ms/step - loss: 0.1220 - acc: 0.9681
Epoch 41/100
25/25 [==============================] - 2s 72ms/step - loss: 0.1345 - acc: 0.9617
Epoch 42/100
25/25 [==============================] - 2s 72ms/step - loss: 0.1173 - acc: 0.9694
Epoch 43/100
25/25 [==============================] - 2s 71ms/step - loss: 0.1083 - acc: 0.9756
Epoch 44/100
25/25 [==============================] - 2s 71ms/step - loss: 0.1089 - acc: 0.9753
Epoch 45/100
25/25 [==============================] - 2s 74ms/step - loss: 0.1124 - acc: 0.9725
Epoch 46/100
25/25 [==============================] - 2s 73ms/step - loss: 0.1247 - acc: 0.9655
Epoch 47/100
25/25 [==============================] -

25/25 [==============================] - 2s 74ms/step - loss: 0.2473 - acc: 0.9175
Epoch 16/100
25/25 [==============================] - 2s 73ms/step - loss: 0.2419 - acc: 0.9172
Epoch 17/100
25/25 [==============================] - 2s 71ms/step - loss: 0.2130 - acc: 0.9363
Epoch 18/100
25/25 [==============================] - 2s 71ms/step - loss: 0.2220 - acc: 0.9245
Epoch 19/100
25/25 [==============================] - 2s 79ms/step - loss: 0.1967 - acc: 0.9395
Epoch 20/100
25/25 [==============================] - 2s 73ms/step - loss: 0.1885 - acc: 0.9445
Epoch 21/100
25/25 [==============================] - 2s 74ms/step - loss: 0.1756 - acc: 0.9479
Epoch 22/100
25/25 [==============================] - 2s 74ms/step - loss: 0.1774 - acc: 0.9482
Epoch 23/100
25/25 [==============================] - 2s 74ms/step - loss: 0.1623 - acc: 0.9544
Epoch 24/100
25/25 [==============================] - 2s 75ms/step - loss: 0.1551 - acc: 0.9545
Epoch 25/100
25/25 [==============================] -

25/25 [==============================] - 2s 73ms/step - loss: 0.0623 - acc: 0.9877
Epoch 1/100
25/25 [==============================] - 5s 73ms/step - loss: 0.6776 - acc: 0.5987
Epoch 2/100
25/25 [==============================] - 2s 80ms/step - loss: 0.6741 - acc: 0.6106
Epoch 3/100
25/25 [==============================] - 2s 74ms/step - loss: 0.6661 - acc: 0.6129
Epoch 4/100
25/25 [==============================] - 2s 75ms/step - loss: 0.6018 - acc: 0.6732
Epoch 5/100
25/25 [==============================] - 2s 73ms/step - loss: 0.5027 - acc: 0.7703
Epoch 6/100
25/25 [==============================] - 2s 74ms/step - loss: 0.4347 - acc: 0.8080
Epoch 7/100
25/25 [==============================] - 2s 73ms/step - loss: 0.4105 - acc: 0.8245
Epoch 8/100
25/25 [==============================] - 2s 73ms/step - loss: 0.3702 - acc: 0.8484
Epoch 9/100
25/25 [==============================] - 2s 74ms/step - loss: 0.3467 - acc: 0.8602
Epoch 10/100
25/25 [==============================] - 2s 80ms/

25/25 [==============================] - 2s 73ms/step - loss: 0.0821 - acc: 0.9832
Epoch 79/100
25/25 [==============================] - 2s 74ms/step - loss: 0.0863 - acc: 0.9799
Epoch 80/100
25/25 [==============================] - 2s 72ms/step - loss: 0.0894 - acc: 0.9788
Epoch 81/100
25/25 [==============================] - 2s 73ms/step - loss: 0.0902 - acc: 0.9790
Epoch 82/100
25/25 [==============================] - 2s 74ms/step - loss: 0.0810 - acc: 0.9833
Epoch 83/100
25/25 [==============================] - 2s 72ms/step - loss: 0.0794 - acc: 0.9825
Epoch 84/100
25/25 [==============================] - 2s 71ms/step - loss: 0.0797 - acc: 0.9830
Epoch 85/100
25/25 [==============================] - 2s 70ms/step - loss: 0.0848 - acc: 0.9808
Epoch 86/100
25/25 [==============================] - 2s 79ms/step - loss: 0.0845 - acc: 0.9814
Epoch 87/100
25/25 [==============================] - 2s 73ms/step - loss: 0.0809 - acc: 0.9827
Epoch 88/100
25/25 [==============================] -

25/25 [==============================] - 2s 73ms/step - loss: 0.1251 - acc: 0.9717
Epoch 57/100
25/25 [==============================] - 2s 71ms/step - loss: 0.1278 - acc: 0.9704
Epoch 58/100
25/25 [==============================] - 2s 74ms/step - loss: 0.1249 - acc: 0.9714
Epoch 59/100
25/25 [==============================] - 2s 72ms/step - loss: 0.1244 - acc: 0.9723
Epoch 60/100
25/25 [==============================] - 2s 73ms/step - loss: 0.1227 - acc: 0.9733
Epoch 61/100
25/25 [==============================] - 2s 72ms/step - loss: 0.1323 - acc: 0.9676
Epoch 62/100
25/25 [==============================] - 2s 73ms/step - loss: 0.1216 - acc: 0.9717
Epoch 63/100
25/25 [==============================] - 2s 73ms/step - loss: 0.1300 - acc: 0.9683
Epoch 64/100
25/25 [==============================] - 2s 73ms/step - loss: 0.1242 - acc: 0.9704
Epoch 65/100
25/25 [==============================] - 2s 73ms/step - loss: 0.1251 - acc: 0.9709
Epoch 66/100
25/25 [==============================] -

25/25 [==============================] - 2s 73ms/step - loss: 0.1336 - acc: 0.9651
Epoch 35/100
25/25 [==============================] - 2s 72ms/step - loss: 0.1271 - acc: 0.9688
Epoch 36/100
25/25 [==============================] - 2s 74ms/step - loss: 0.1328 - acc: 0.9662
Epoch 37/100
25/25 [==============================] - 2s 73ms/step - loss: 0.1277 - acc: 0.9689
Epoch 38/100
25/25 [==============================] - 2s 73ms/step - loss: 0.1234 - acc: 0.9715
Epoch 39/100
25/25 [==============================] - 2s 75ms/step - loss: 0.1285 - acc: 0.9676
Epoch 40/100
25/25 [==============================] - 2s 74ms/step - loss: 0.1427 - acc: 0.9602
Epoch 41/100
25/25 [==============================] - 2s 75ms/step - loss: 0.1360 - acc: 0.9654
Epoch 42/100
25/25 [==============================] - 2s 73ms/step - loss: 0.1196 - acc: 0.9720
Epoch 43/100
25/25 [==============================] - 2s 74ms/step - loss: 0.1194 - acc: 0.9723
Epoch 44/100
25/25 [==============================] -

25/25 [==============================] - 2s 74ms/step - loss: 0.3028 - acc: 0.8971
Epoch 13/100
25/25 [==============================] - 2s 74ms/step - loss: 0.2802 - acc: 0.9086
Epoch 14/100
25/25 [==============================] - 2s 73ms/step - loss: 0.2760 - acc: 0.9080
Epoch 15/100
25/25 [==============================] - 2s 73ms/step - loss: 0.2544 - acc: 0.9196
Epoch 16/100
25/25 [==============================] - 2s 73ms/step - loss: 0.2546 - acc: 0.9160
Epoch 17/100
25/25 [==============================] - 2s 74ms/step - loss: 0.2430 - acc: 0.9266
Epoch 18/100
25/25 [==============================] - 2s 75ms/step - loss: 0.2400 - acc: 0.9245
Epoch 19/100
25/25 [==============================] - 2s 73ms/step - loss: 0.2143 - acc: 0.9377
Epoch 20/100
25/25 [==============================] - 2s 73ms/step - loss: 0.2160 - acc: 0.9368
Epoch 21/100
25/25 [==============================] - 2s 74ms/step - loss: 0.2023 - acc: 0.9452
Epoch 22/100
25/25 [==============================] -

25/25 [==============================] - 2s 73ms/step - loss: 0.0946 - acc: 0.9790
Epoch 98/100
25/25 [==============================] - 2s 73ms/step - loss: 0.0952 - acc: 0.9793
Epoch 99/100
25/25 [==============================] - 2s 73ms/step - loss: 0.0911 - acc: 0.9811
Epoch 100/100
25/25 [==============================] - 2s 72ms/step - loss: 0.0939 - acc: 0.9812
Epoch 1/100
25/25 [==============================] - 5s 74ms/step - loss: 0.6782 - acc: 0.5896
Epoch 2/100
25/25 [==============================] - 2s 75ms/step - loss: 0.6725 - acc: 0.6081
Epoch 3/100
25/25 [==============================] - 2s 75ms/step - loss: 0.6666 - acc: 0.6072
Epoch 4/100
25/25 [==============================] - 2s 72ms/step - loss: 0.6138 - acc: 0.6585
Epoch 5/100
25/25 [==============================] - 2s 74ms/step - loss: 0.5165 - acc: 0.7608
Epoch 6/100
25/25 [==============================] - 2s 75ms/step - loss: 0.4443 - acc: 0.8046
Epoch 7/100
25/25 [==============================] - 2s 73

25/25 [==============================] - 2s 71ms/step - loss: 0.0879 - acc: 0.9804
Epoch 76/100
25/25 [==============================] - 2s 72ms/step - loss: 0.0853 - acc: 0.9814
Epoch 77/100
25/25 [==============================] - 2s 72ms/step - loss: 0.0842 - acc: 0.9812
Epoch 78/100
25/25 [==============================] - 2s 71ms/step - loss: 0.0836 - acc: 0.9825
Epoch 79/100
25/25 [==============================] - 2s 72ms/step - loss: 0.0917 - acc: 0.9780
Epoch 80/100
25/25 [==============================] - 2s 71ms/step - loss: 0.0811 - acc: 0.9820
Epoch 81/100
25/25 [==============================] - 2s 71ms/step - loss: 0.0855 - acc: 0.9819
Epoch 82/100
25/25 [==============================] - 2s 73ms/step - loss: 0.0856 - acc: 0.9812
Epoch 83/100
25/25 [==============================] - 2s 72ms/step - loss: 0.0882 - acc: 0.9804
Epoch 84/100
25/25 [==============================] - 2s 72ms/step - loss: 0.0858 - acc: 0.9814
Epoch 85/100
25/25 [==============================] -

25/25 [==============================] - 2s 70ms/step - loss: 0.0927 - acc: 0.9801
Epoch 54/100
25/25 [==============================] - 2s 69ms/step - loss: 0.0912 - acc: 0.9796
Epoch 55/100
25/25 [==============================] - 2s 71ms/step - loss: 0.0904 - acc: 0.9809
Epoch 56/100
25/25 [==============================] - 2s 71ms/step - loss: 0.0874 - acc: 0.9804
Epoch 57/100
25/25 [==============================] - 2s 70ms/step - loss: 0.0936 - acc: 0.9762
Epoch 58/100
25/25 [==============================] - 2s 70ms/step - loss: 0.0914 - acc: 0.9795
Epoch 59/100
25/25 [==============================] - 2s 70ms/step - loss: 0.0931 - acc: 0.9782
Epoch 60/100
25/25 [==============================] - 2s 70ms/step - loss: 0.1081 - acc: 0.9730
Epoch 61/100
25/25 [==============================] - 2s 70ms/step - loss: 0.0922 - acc: 0.9778
Epoch 62/100
25/25 [==============================] - 2s 69ms/step - loss: 0.0932 - acc: 0.9780
Epoch 63/100
25/25 [==============================] -

25/25 [==============================] - 2s 72ms/step - loss: 0.1304 - acc: 0.9665
Epoch 32/100
25/25 [==============================] - 2s 73ms/step - loss: 0.1306 - acc: 0.9647
Epoch 33/100
25/25 [==============================] - 2s 72ms/step - loss: 0.1337 - acc: 0.9617
Epoch 34/100
25/25 [==============================] - 2s 72ms/step - loss: 0.1354 - acc: 0.9621
Epoch 35/100
25/25 [==============================] - 2s 72ms/step - loss: 0.1170 - acc: 0.9722
Epoch 36/100
25/25 [==============================] - 2s 74ms/step - loss: 0.1147 - acc: 0.9719
Epoch 37/100
25/25 [==============================] - 2s 73ms/step - loss: 0.1103 - acc: 0.9746
Epoch 38/100
25/25 [==============================] - 2s 73ms/step - loss: 0.1167 - acc: 0.9710
Epoch 39/100
25/25 [==============================] - 2s 73ms/step - loss: 0.1148 - acc: 0.9723
Epoch 40/100
25/25 [==============================] - 2s 72ms/step - loss: 0.1089 - acc: 0.9740
Epoch 41/100
25/25 [==============================] -

In [39]:
from sklearn.metrics import accuracy_score
from sklearn.metrics import precision_score
from sklearn.metrics import recall_score
from sklearn.metrics import f1_score
from sklearn.metrics import roc_auc_score
preds = final_model.predict(X_test)

precision = precision_score(y_test, preds)
recall = recall_score(y_test, preds)
f1 = f1_score(y_test, preds)
roc_auc = roc_auc_score(y_test, preds)
acc = accuracy_score(y_test, preds)

print(f' accuracy : {acc}, precision : {precision}, recall : {recall}, f1 : {f1}, roc_auc : {roc_auc}')

 accuracy : 0.7436893203883496, precision : 0.7903930131004366, recall : 0.7801724137931034, f1 : 0.7852494577006508, roc_auc : 0.7344946347733751


**KerasClassifier(build_fn=lambda:get_model(), epochs=100, batch_size=256)**
- 1개 : 0.7411003236245954
- 5개 : 0.7443365695792881
- 15개 : 0.7417475728155339
- 20개 : 0.7411003236245954
- 30개 : 
- 50개 : {accuracy : 0.7436893203883496, precision : 0.7903930131004366, recall : 0.7801724137931034, f1 : 0.7852494577006508, roc_auc : 0.7344946347733751}

In [32]:
%%time
tf.random.set_seed(seed_num)

def get_model(nums):
    lstm = Sequential()
    lstm.add(InputLayer(input_shape=(x.shape[1],x.shape[2])))
    lstm.add(LSTM(units=nums[0], activation='hard_sigmoid', return_sequences=True))
    lstm.add(LSTM(units=nums[1], activation='hard_sigmoid', return_sequences=True))
    lstm.add(Dropout(0.2))
    lstm.add(LSTM(units=nums[2], activation='hard_sigmoid', return_sequences=True))
    lstm.add(LSTM(units=nums[3], activation='hard_sigmoid', return_sequences=False))
    lstm.add(Dropout(0.2))
    lstm.add(Dense(units=1, activation='sigmoid'))

    lstm.compile(optimizer= keras.optimizers.Adam(learning_rate = 0.001), 
                          loss = "binary_crossentropy", metrics=['acc'])
    return lstm

# 서로 다른 모델을 만들어 합치기 (Model Ensemble)
model1 = KerasClassifier(build_fn=lambda:get_model(nums = [256, 128, 64, 32]), epochs=50, batch_size=256)
model1._estimator_type="classifier"
model2 = KerasClassifier(build_fn=lambda:get_model(nums = [128, 128, 64, 32]), epochs=50, batch_size=256)
model2._estimator_type="classifier"
model3 = KerasClassifier(build_fn=lambda:get_model(nums = [128, 64, 64, 32]), epochs=50, batch_size=256) 
model3._estimator_type="classifier"

estimator = [('model1', model1), ('model2', model2), ('model3', model3)]

final_model = VotingClassifier(estimators = estimator, voting = 'soft')
final_model.fit(X_train,y_train)

preds = final_model.predict(X_test)

preds[preds>0.5]=1
preds[preds<=0.5]=0

from sklearn import metrics 
print('정확도 :', metrics.accuracy_score(y_test, preds))
print("RMSE : ", RMSE(y_test, preds))

Epoch 1/50
25/25 [==============================] - 4s 67ms/step - loss: 0.6792 - acc: 0.5993
Epoch 2/50
25/25 [==============================] - 2s 66ms/step - loss: 0.6709 - acc: 0.6105
Epoch 3/50
25/25 [==============================] - 2s 66ms/step - loss: 0.6374 - acc: 0.6276
Epoch 4/50
25/25 [==============================] - 2s 67ms/step - loss: 0.5331 - acc: 0.7401
Epoch 5/50
25/25 [==============================] - 2s 66ms/step - loss: 0.4549 - acc: 0.8017
Epoch 6/50
25/25 [==============================] - 2s 67ms/step - loss: 0.4073 - acc: 0.8255
Epoch 7/50
25/25 [==============================] - 2s 66ms/step - loss: 0.3889 - acc: 0.8348
Epoch 8/50
25/25 [==============================] - 2s 67ms/step - loss: 0.3682 - acc: 0.8468
Epoch 9/50
25/25 [==============================] - 2s 66ms/step - loss: 0.3312 - acc: 0.8711
Epoch 10/50
25/25 [==============================] - 2s 66ms/step - loss: 0.3098 - acc: 0.8822
Epoch 11/50
25/25 [==============================] - 2s 66m

25/25 [==============================] - 2s 67ms/step - loss: 0.1752 - acc: 0.9489
Epoch 23/50
25/25 [==============================] - 2s 66ms/step - loss: 0.1659 - acc: 0.9523
Epoch 24/50
25/25 [==============================] - 2s 66ms/step - loss: 0.1548 - acc: 0.9588
Epoch 25/50
25/25 [==============================] - 2s 66ms/step - loss: 0.1439 - acc: 0.9604
Epoch 26/50
25/25 [==============================] - 2s 67ms/step - loss: 0.1470 - acc: 0.9594
Epoch 27/50
25/25 [==============================] - 2s 67ms/step - loss: 0.1367 - acc: 0.9636
Epoch 28/50
25/25 [==============================] - 2s 67ms/step - loss: 0.1426 - acc: 0.9621
Epoch 29/50
25/25 [==============================] - 2s 68ms/step - loss: 0.1324 - acc: 0.9651
Epoch 30/50
25/25 [==============================] - 2s 66ms/step - loss: 0.1294 - acc: 0.9655
Epoch 31/50
25/25 [==============================] - 2s 66ms/step - loss: 0.1275 - acc: 0.9672
Epoch 32/50
25/25 [==============================] - 2s 66ms/s

25/25 [==============================] - 2s 66ms/step - loss: 0.1150 - acc: 0.9743
정확도 : 0.7430420711974111
RMSE :  5.07e-01
CPU times: user 10min 34s, sys: 13min 40s, total: 24min 15s
Wall time: 4min 21s


In [ ]:
#https://machinelearningmastery.com/use-keras-deep-learning-models-scikit-learn-python/
from sklearn.model_selection import GridSearchCV

# Function to create model, required for KerasClassifier
def get_model(init='glorot_uniform'):
    lstm = Sequential()
    lstm.add(InputLayer(input_shape=(x.shape[1],x.shape[2])))
    lstm.add(LSTM(units=128, activation='hard_sigmoid', kernel_initializer=init, return_sequences=True))
    lstm.add(LSTM(units=64, activation='hard_sigmoid', kernel_initializer=init, return_sequences=True))
    lstm.add(Dropout(0.2))
    lstm.add(LSTM(units=64, activation='hard_sigmoid', kernel_initializer=init, return_sequences=True))
    lstm.add(LSTM(units=32, activation='hard_sigmoid', kernel_initializer=init, return_sequences=False))
    lstm.add(Dropout(0.2))
    lstm.add(Dense(units=1, activation='sigmoid'))

    lstm.compile(optimizer= keras.optimizers.Adam(learning_rate = 0.001), 
                          loss = "binary_crossentropy", metrics=['acc'])
    return lstm

# fix random seed for reproducibility
seed_num = 42
random.seed(seed_num)

x = np.load('/project/LSH/x_(7727,10,4068).npy')
y = np.load('/project/LSH/y_(7727,1).npy')

idx = list(range(len(x)))
random.shuffle(idx)

i = round(x.shape[0]*0.8)
X_train, y_train = x[idx[:i],:,:], y[idx[:i]]
X_test, y_test = x[idx[i:],:,:], y[idx[i:]]

# create model
model = KerasClassifier(build_fn=create_model, verbose=0)

# grid search epochs, batch size and optimizer
init = ['glorot_uniform', 'normal', 'uniform']
epochs = [50, 100, 150]
batches = [8, 16, 32, 64]
param_grid = dict(optimizer=optimizers, epochs=epochs, batch_size=batches, init=init)

grid = GridSearchCV(estimator=model, param_grid=param_grid)
grid_result = grid.fit(X_train, y_train)

# summarize results
print("Best: %f using %s" % (grid_result.best_score_, grid_result.best_params_))

means = grid_result.cv_results_['mean_test_score']
stds = grid_result.cv_results_['std_test_score']
params = grid_result.cv_results_['params']

for mean, stdev, param in zip(means, stds, params):
    print("%f (%f) with: %r" % (mean, stdev, param))